keps/20200901-standard-crd.md

```diff
@@ -0,0 +1,240 @@
+4. AS a KES user i want to fetch **from multiple** stores and store the secrets **in a single** Frontend
+5. AS a KES operator i want to limit the access to certain stores or subresources (e.g. having one central KES instance that handles all ES - similar to `iam.amazonaws.com/permitted` annotation per namespace)
+
+### Stores
+
+These stores are relevant:
+* AWS Secure Systems Manager Parameter Store
+* AWS Secrets Manager
+* Hashicorp Vault
+* Azure Key Vault
+* Alibaba Cloud KMS Secret Manager
+* Google Cloud Platform Secret Manager
+* Kubernetes (see #422)
+* noop (see #476)
+
+### Frontends
+
+* Kind=Secret
+* Kind=ConfigMap
+* *potentially* we could sync store to store
+
+## Proposal
+
+### API
+
+### External Secret
+
+The `ExternalSecret` CustomResourceDefinition is **namespaced**. It defines the following:
+1. source for the secret (store)
+2. sink for the secret (fronted)
+3. and a mapping to translate the keys
+
+```yaml
+apiVersion: kes.io/v1alpha1
+kind: ExternalSecret
+metadata: {...}
+
+spec:
+
+  # optional.
+  # used to select the correct KES instance (think: ingress.ingressClassName)
+  # There is no need for a indirection (e.g. having a extra resource like kind=IngressClass)
+  # the KES controller is instantiated with a specific class name
+  # and filters ES based on this property
+  className: "dev"
+
+  # the amount of time before the values will be read again from the store
+  refreshInterval: "1h"
+
+  # there can only be one frontend per ES
+  # this is the "thing" that is created by KES.
+  # conceptually speaking a frontend is just a thing that can hold k/v pairs
+  frontend:
+    secret:
+      # do we need a api version here? who handles upgrades?
+      apiVersion: v1
+      name: my-secret
+      template:
+        type: kubernetes.io/dockerconfigjson # or TLS...
+        metadata:
+          annotations: {}
+          labels: {}
+    # of course only one frontend should be defined
+    configMap:
+      # do we need a api version here? who handles upgrades?
+      apiVersion: v1 #
+      name: my-configmap
+      template:
+        metadata:
+          labels:
+            app: foo
+
+  data:
+
+    # EXAMPLE 1: simple mapping
+    # one key from a store may hold multiple values
+    # we need a way to map the values to the frontend
+    # it is the responsibility of the store implementation to know how to extract a value
+  - key: /corp.org/dev/certs/ingress
+    property: pubcert
+    name: tls.crt
+  - key: /corp.org/dev/certs/ingress
+    property: privkey
+    name: tls.key
+
+  # EXAMPLE 2: multiple stores per ES
+  # we also need a way to fetch secrets from multiple stores
+  # thus, we need a way to define a store per data key
+  - key: /rds/database-secret
+    name: database
+    storeRef:
+      name: foo
+  - key: /users/my-db-user
+    name: username
+    storeRef:
+      name: bar
```

Comment:
I like this syntax, `data` could be a map of key references while `dataFrom` could be a list of key references.

```yaml
data:
  tls.crt:
    key: /corp.org/dev/certs/ingress
    property: pubcert
  tls.key:
    key: /corp.org/dev/certs/ingress
    property: privkey

dataFrom:
- key: /rds/database-secret
```

Reply:
+1 for the data block, but im not sure that the store reference belongs in data

Reply:
> I like this syntax, data could be a map of key references while dataFrom could be a list of key references.

:+1: @mcavoyk This is just for maintaining the correct order, right? 

> im not sure that the store reference belongs in data

@dirtycajunrice We need to merge multiple secrets from $CLOUDPROVIDER into a single secret. E.g. Alertmanager needs multiple API Keys.

Reply:
@moolen 
> +1 @mcavoyk This is just for maintaining the correct order, right?
`dataFrom` was designed to be a way to use all properties from a key as is, without having to specify them individually.
> @DirtyCajunRice We need to merge multiple secrets from $CLOUDPROVIDER into a single secret. E.g. Alertmanager needs multiple API Keys.
Ill reply to that one there :smile:

Reply:
>👍 @mcavoyk This is just for maintaining the correct order, right?

> dataFrom was designed to be a way to use all properties from a key as is, without having to specify them individually.

Yepp, in KES `dataFrom` is expected to be a JSON payload. Each property from the object will be set as a field in the generated secrets data. If there are multiple sources with the same resulting field the last source in the `dataFrom` that defines that field will be used. And then if the same field is set in the external secrets `data` that will override any `dataFrom` field.

For `data` the ordering doesn't matter as the name would be unique anyway. So it would just be silly to duplicate to use the same 😄

Reply:
same as above: thank you for your input! I integrated your changes into the spec. Feel free to un-resolve if something is missing or wrong.
Changelog:
* using data object syntax in favor of array
* add meta info for implementation specifics